pkg/volume/util/operationexecutor/operation_executor.go

```diff
@@ -607,6 +578,54 @@ func (oe *operationExecutor) generateDetachVolumeFunc(
 	}, nil
 }
 
+func (oe *operationExecutor) verifyVolumeIsSafeToDetach(
+	volumeToDetach AttachedVolume) error {
+	// Fetch current node object
+	node, fetchErr := oe.kubeClient.Core().Nodes().Get(volumeToDetach.NodeName)
+	if fetchErr != nil {
+		if errors.IsNotFound(fetchErr) {
+			glog.Warningf("Node %q not found on API server. DetachVolume will skip safe to detach check.",
+				volumeToDetach.NodeName,
+				volumeToDetach.VolumeName,
+				volumeToDetach.VolumeSpec.Name())
+			return nil
```

Comment:
> By returning nil here, it can unblock DetachVolume in operation_executor. 

Returning nil will not block the operation_executor, returning an error will.

> 1. If the volume is still attached by another node, attach will fail. 

Correct.

> 2. If the node to which the volume is originally attached is removed from API server (by kubectl delete), attach failed at first but succeed eventually because volume is detached after timeout (waiting for unmount status update). In this case verifySafeToDetach is set to false

Detach from the original node would fail because of the bug this PR is fixing. After this PR, what you stated will be correct.

> 3. If the node to which the volume is originally attached is deleted (by gcloud command), attach will succeed? I am not quite sure in this situation verifySafeToDetach is set to true or false in reconciler.

Yes, volume will be detached thanks to PR #29485, and be ok to reattach to another node.

> 4. If the previous operation on the volume failed, it will block on backoff

Not sure what you mean here.

> So it is not clear to me that in the issue #29903, during node upgrading, if the old node is deleted, the following attach to the new node should succeed. Also from the kubelet log I checked, kubelet verifyVolumeAttached succeed because the volume is in the attached list of the new node, but failed on checking the devicePath before trying to mount volume. I am thinking there might be some race condition issue in the attachdetach controller

The issue was that when the cluster is updated the old machine is deleted/removed from API server. Then the master would 1) unable to correctly update the list of attached volumes and 2) not be able to detach the volume with the node obj missing from the API server. This PR fixes those bugs.